### PR TITLE
Revive sync

### LINKED ISSUE #31: Add batch processors
Upon receiving a new batch that wasn't submitted by us we want to parse calldata, process all transactions and store them in our DB

https://github.com/thehubbleproject/hubble-commander/blob/master/core/bazooka.go#L220

### PR DESCRIPTION
**Pull request by vaibhavchellani**
Fixes #31

### CODE REVIEW

**Comment by vaibhavchellani** [comment]
State validation during sync is not complete, opened a new issues for signature validation: #74

File: core/batch.go [hunk 1]
```diff
@@ -48,6 +59,36 @@ func (db *DB) GetBatchByIndex(index uint64) (batch Batch, err error) {
...
 
-func (db *DB) CommitBatch(batch Batch) error {
-	return db.Instance.Update(batch).Error
+func (db *DB) CommitBatch(ID uint64) error {
+	batch, err := db.GetBatchByIndex(ID)
+	if err != nil {
+		return err
+	}
+
+	if batch.Status != BATCH_BROADCASTED && batch.Status == BATCH_COMMITTED {
```
Comment (ChihChengLiang) [inline core/batch.go:68]: I think we can safely remove `batch.Status != BATCH_BROADCASTED`. 
- When `batch.Status != BATCH_COMMITTED` the whole condition evaluates false
- when `batch.Status == BATCH_COMMITTED`, `batch.Status != BATCH_BROADCASTED` evaluates true, so the whole condition evaluates true too.

```suggestion
	if batch.Status == BATCH_COMMITTED {
```

File: core/tx.go [hunk 2]
```diff
@@ -260,3 +271,79 @@ func (db *DB) FetchMPWithID(id uint64, accountMP *StateMerkleProof) (err error)
...
+func (tx *Tx) Validate(bz Bazooka, currentRoot ByteArray) (newRoot ByteArray, err error) {
+	fromStateProof, toStateProof, txDBConn, err := tx.GetVerificationData()
+	if err != nil {
+		return
+	}
+
+	newRoot, err = bz.ProcessTx(currentRoot, *tx, fromStateProof, toStateProof)
+	if err != nil {
+		if txDBConn.Instance != nil {
+			txDBConn.Instance.Rollback()
```
Comment (ChihChengLiang) [inline core/tx.go:285]: What is rollbacked here?

File: core/tx.go [hunk 3]
```diff
@@ -260,3 +271,79 @@ func (db *DB) FetchMPWithID(id uint64, accountMP *StateMerkleProof) (err error)
...
+			txDBConn.Close()
+		}
+		return
+	}
+	if txDBConn.Instance != nil {
+		txDBConn.Instance.Commit()
```
Comment (ChihChengLiang) [inline core/tx.go:291]: What insert or update is committed here?

File: core/tx.go [hunk 4]
```diff
@@ -260,3 +271,79 @@ func (db *DB) FetchMPWithID(id uint64, accountMP *StateMerkleProof) (err error)
...
+		newRoot, err := tx.Validate(LoadedBazooka, currentRoot)
+		if err != nil {
+			return commitments, err
+		}
+		if i%COMMITMENT_SIZE == 0 {
+			txInCommitment := txs[i : i+COMMITMENT_SIZE]
+			aggregatedSig, err := aggregateSignatures(txInCommitment)
+			if err != nil {
+				if isSyncing && err == ErrSignatureNotPresent {
+					continue
```
Comment (ChihChengLiang) [inline core/tx.go:321]: Does the `continue` here mean the line `commitments = append(commitments, commitment)` is skipped?

File: core/bazooka.go [hunk 5]
```diff
@@ -251,36 +289,53 @@ func (b *Bazooka) applyTransferTx(sender, receiver []byte, tx Tx) (updatedSender
 	return updates.NewSender, updates.NewReceiver, nil
 }
 
-func (b *Bazooka) applyMassMigrationTx(sender, receiver []byte, tx Tx) (updatedSender, updatedReceiver []byte, err error) {
-	opts := bind.CallOpts{From: config.OperatorAddress}
-	updates, err := b.Frontend.ValidateAndApplyMassMigration(&opts, sender, tx.Data)
-	if err != nil {
-		return
-	}
+// func (b *Bazooka) applyMassMigrationTx(sender, receiver []byte, tx Tx) (updatedSender, updatedReceiver []byte, err error) {
```
Comment (ChihChengLiang) [inline core/bazooka.go:292]: nitpick: For these areas of commented off code, it's fine to just remove them.

**Review by ChihChengLiang — APPROVED** [review]
Great work! I left comments on minor issues to address, looks good to me in general!

File: core/tx.go (hunk 3 above)
Comment (vaibhavchellani) [reply core/tx.go:291]: There are updates performed in GetTxVerificationData() via a MySQL transaction, we can rollback those updates if processTx is invalid.

File: core/tx.go (hunk 2 above)
Comment (vaibhavchellani) [reply core/tx.go:285]: If processTx is invalid we can rollback the changes made while witness creation

File: core/tx.go (hunk 4 above)
Comment (vaibhavchellani) [reply core/tx.go:321]: No it goes there, it simply means that if we are syncing and tx.Sig is nil we don't have to bubble up error as that is expected. I do think this piece is smelly and should be refactored in the future.

File: core/batch.go (hunk 1 above)
Comment (vaibhavchellani) [reply core/batch.go:68]: Yeah I think your suggestion assumes that the status can be either BATCH_BROADCASTED or BATCH_COMMITTED while I was trying to protect from invalid status codes too, which I think is stupid🤐

Comment (vaibhavchellani) [reply core/batch.go:68]: Fixing

**Pull request merged**